staging/src/k8s.io/client-go/util/consistencydetector/list_data_consistency_detector.go

```diff
@@ -0,0 +1,82 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package consistencydetector
+
+import (
+	"context"
+	"os"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var dataConsistencyDetectionForListAgainstCacheEnabled = false
+
+func init() {
+	dataConsistencyDetectionForListAgainstCacheEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_LIST_AGAINST_CACHE_INCONSISTENCY_DETECTOR"))
+}
+
+// CheckListAgainstCacheDataConsistencyIfRequested performs a data consistency check only when
+// the KUBE_LIST_AGAINST_CACHE_INCONSISTENCY_DETECTOR environment variable was set during a binary startup
+// for requests served from the watch-cache.
+//
+// The consistency check is meant to be enforced only in the CI, not in production.
+// The check ensures that data retrieved by a list api call against the watch-cache
+// is exactly the same as data received by the list api call against etcd.
+//
+// Note that this function will panic when data inconsistency is detected.
+// This is intentional because we want to catch it in the CI.
+func CheckListAgainstCacheDataConsistencyIfRequested[T runtime.Object](ctx context.Context, identity string, listItemsFn ListFunc[T], optionsUsedToReceiveList metav1.ListOptions, receivedList runtime.Object) {
+	if !dataConsistencyDetectionForListAgainstCacheEnabled || wasListRequestServedFromStorage(optionsUsedToReceiveList) {
+		return
+	}
+	receivedListMeta, err := meta.ListAccessor(receivedList)
+	if err != nil {
+		panic(err)
+	}
+	rawListItems, err := meta.ExtractListWithAlloc(receivedList)
+	if err != nil {
+		panic(err) // this should never happen
+	}
+	lastSyncedResourceVersion := receivedListMeta.GetResourceVersion()
+	CheckDataConsistency(ctx, identity, lastSyncedResourceVersion, listItemsFn, optionsUsedToReceiveList, func() []runtime.Object { return rawListItems })
+}
+
+// wasListRequestServedFromStorage based on the passed ListOptions determines
+// if the request was served from the underlying storage (instead of the watch-cache).
+//
+// Note that this function is unable to detect the following cases:
+//
+// if ResourceVersion = "" and ConsistendListFromCache is disabled or RequestWatchProgress isn't supported,
+// then the request will be served from the storage.
+func wasListRequestServedFromStorage(opts metav1.ListOptions) bool {
+	resourceVersion := opts.ResourceVersion
+	match := opts.ResourceVersionMatch
+
+	// watch-cache doesn't support continuations, so serve them from etcd.
+	hasContinuation := len(opts.Continue) > 0
+	// TODO: clarify rows 1 and 3 from https://github.com/kubernetes/enhancements/blob/9f92967516301bbd564b7db221fe92ce3a253792/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#what-if-the-watch-cache-is-stale
```

Comment:
@wojtek-t could you take a look at rows 1 and 3 from the KEP ? - I think that the cacher doesn't support `Limit=N` in these cases.

Reply:
It doesn't - it's delegating do etcd:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L739

That said, we can easily implement it [and probably should, though separately]

Reply:
OK, how do we want to capture it ? I can create an issue, would that be ok ?

Reply:
yes - let's create a separate issue for that